### ISSUE TITLE
Failed to compute shorthash for libnvrtc.so when compiling application with libtorch 1.8.0

### ISSUE THREAD

**Issue by lamhoangtung** [pytorch/pytorch#53350]
## 🐛 Bug

When building an application with libtorch 1.8.0, I got this warning:
```console
CMake Warning at build/libtorch/share/cmake/Caffe2/public/cuda.cmake:198 (message):
  Failed to compute shorthash for libnvrtc.so
Call Stack (most recent call first):
  build/libtorch/share/cmake/Caffe2/Caffe2Config.cmake:88 (include)
  build/libtorch/share/cmake/Torch/TorchConfig.cmake:68 (find_package)
  CMakeLists.txt:109 (find_package)
```
The previous version 1.7.1 had no such warning. 

## To Reproduce

I had created a simple project to reproduce this: https://github.com/lamhoangtung/dummy-libtorch

To use it, run this on a machine with CUDA or the docker image `nvidia/cuda:11.1.1-cudnn8-devel-ubuntu20.04`:
```bash
git clone https://github.com/lamhoangtung/dummy-libtorch
cd dummy-libtorch
sudo apt update
sudo apt install unzip cmake
sh download_libtorch.sh
sh build.sh
```

The output shoud be like this:
```console
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found CUDA: /usr/local/cuda-11.1 (found version "11.1")
-- Caffe2: CUDA detected: 11.1
-- Caffe2: CUDA nvcc is: /usr/local/cuda-11.1/bin/nvcc
-- Caffe2: CUDA toolkit directory: /usr/local/cuda-11.1
-- Caffe2: Header version is: 11.1
-- Found CUDNN: /usr/local/cuda-11.1/lib64/libcudnn.so
-- Found cuDNN: v8.1.1  (include: /usr/local/cuda-11.1/include, library: /usr/local/cuda-11.1/lib64/libcudnn.so)
CMake Warning at libtorch/share/cmake/Caffe2/public/cuda.cmake:198 (message):
  Failed to compute shorthash for libnvrtc.so
Call Stack (most recent call first):
  libtorch/share/cmake/Caffe2/Caffe2Config.cmake:88 (include)
  libtorch/share/cmake/Torch/TorchConfig.cmake:68 (find_package)
  CMakeLists.txt:4 (find_package)


-- Autodetected CUDA architecture(s):  8.6 6.1
-- Added CUDA NVCC flags for: -gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_61,code=sm_61
-- Found Torch: /home/techainer/linus/dummy-libtorch/libtorch/lib/libtorch.so
-- Configuring done
-- Generating done
-- Build files have been written to: /home/techainer/linus/dummy-libtorch/build
Scanning dependencies of target simple_model
[ 50%] Building CXX object CMakeFiles/simple_model.dir/model.cpp.o
[100%] Linking CXX executable simple_model
[100%] Built target simple_model
```

## Expected behavior

No such warning above

## Environment

My environment was init from this docker images: `nvidia/cuda:11.1.1-cudnn8-devel-ubuntu20.04` 

```
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.19.5

Python version: 3.8 (64-bit runtime)
Is CUDA available: N/A
CUDA runtime version: 11.1.105
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce RTX 3090

Nvidia driver version: 460.56
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect
```

## Additional context

After doing some research, I found the warning will be raise from this: [cuda.cmake](https://github.com/pytorch/pytorch/blob/v1.8.0/cmake/public/cuda.cmake) file.

```cmake
 execute_process(
    COMMAND "${PYTHON_EXECUTABLE}" -c
    "import hashlib;hash=hashlib.sha256();hash.update(open('${CUDA_NVRTC_LIB}','rb').read());print(hash.hexdigest()[:8])"
    RESULT_VARIABLE _retval
    OUTPUT_VARIABLE CUDA_NVRTC_SHORTHASH)
  if(NOT _retval EQUAL 0)
    message(WARNING "Failed to compute shorthash for libnvrtc.so")
    set(CUDA_NVRTC_SHORTHASH "XXXXXXXX")
```

Which mean that the python script: `"import hashlib;hash=hashlib.sha256();hash.update(open('${CUDA_NVRTC_LIB}','rb').read());print(hash.hexdigest()[:8])"` had failed.
```console
Traceback (most recent call last):
  File "<string>", line 1, in <module>
FileNotFoundError: [Errno 2] No such file or directory: ''
```

I suspect that my shell does not had the variable `CUDA_NVRTC_LIB` configure properly. So I tried to find it, and set it to this according to this [documentation](https://docs.nvidia.com/cuda/nvrtc/index.html):
```bash
export CUDA_NVRTC_LIB=/usr/local/cuda-11.1/include/nvrtc.h
```

And then rerun the python script above and it seems to work with an output like this:
```console
703109d3
```

But the problem `Failed to compute shorthash for libnvrtc.so` still persist when I run cmake build

cc @malfet @seemethere @walterddr

**Comment by lamhoangtung** [comment]
@christian-rauch My executable runs with CUDA without issues too but the warning make me feel uncomfortable :v

**Comment by alantess** [comment]
I'm having the same problem as well on my new machine. My executable won't run either.  Pretty sure, it has to do with the hashing

**Comment by malfet** [comment]
Some clarification: this codepath is only relevant if one wants to build wheel with CUDA dependencies embedded into it.
That said, I would like to understand why this check fails.
Can someone please re-run cmake in build folder as `nohup cmake --trace-expand .` and share nohup.out somewhere?
This is how relevant portion looks like on my devbox:
```
/home/nshulga/git/pytorch/cmake/public/cuda.cmake(188):  find_library(CUDA_NVRTC_LIB nvrtc PATHS /usr/local/cuda-10.2 PATH_SUFFIXES lib lib64 lib/x64 )
/home/nshulga/git/pytorch/cmake/public/cuda.cmake(191):  if(CUDA_NVRTC_LIB AND NOT CUDA_NVRTC_SHORTHASH )
/home/nshulga/git/pytorch/cmake/public/cuda.cmake(192):  execute_process(COMMAND /usr/bin/python -c import hashlib;hash=hashlib.sha256();hash.update(open('/usr/local/cuda-10.2/lib64/libnvrtc.so','rb').read());print(hash.hexdigest()[:8]) RESULT_VARIABLE _retval OUTPUT_VARIABLE CUDA_NVRTC_SHORTHASH )
/home/nshulga/git/pytorch/cmake/public/cuda.cmake(197):  if(NOT _retval EQUAL 0 )
/home/nshulga/git/pytorch/cmake/public/cuda.cmake(200):  else()
/home/nshulga/git/pytorch/cmake/public/cuda.cmake(201):  string(STRIP 08c4863f
 CUDA_NVRTC_SHORTHASH )
/home/nshulga/git/pytorch/cmake/public/cuda.cmake(202):  message(STATUS /usr/local/cuda-10.2/lib64/libnvrtc.so shorthash is 08c4863f )
-- /usr/local/cuda-10.2/lib64/libnvrtc.so shorthash is 08c4863f

```

P.S. CUDA_NVRTC_LIB is not an environment but a CMAKE variable, which should be defined in `build/CMakeCache.txt`

**Comment by lamhoangtung** [comment]
@malfet. Here is my nohup.out file
https://drive.google.com/file/d/1SG2eAUyfI1xysorYLWCi-iGFiLajHXv5/view?usp=sharing

**Comment by lamhoangtung** [comment]
Also I had tried to set `CUDA_NVRTC_LIB` as a CMAKE variable like this:
```cmake
cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
project(simple_model)

set(CUDA_NVRTC_LIB "/usr/local/cuda-11.1/include/nvrtc.h")
find_package( Torch REQUIRED )

add_executable(simple_model model.cpp)
target_link_libraries(simple_model "${TORCH_LIBRARIES}")
set_property(TARGET simple_model PROPERTY CXX_STANDARD 14)
```

But the problem still persist:
```
CMake Warning at libtorch/share/cmake/Caffe2/public/cuda.cmake:198 (message):
  Failed to compute shorthash for libnvrtc.so
Call Stack (most recent call first):
  libtorch/share/cmake/Caffe2/Caffe2Config.cmake:88 (include)
  libtorch/share/cmake/Torch/TorchConfig.cmake:68 (find_package)
  CMakeLists.txt:5 (find_package)
```

Also here is the nohup.out file after setting the `CUDA_NVRTC_LIB` as a CMAKE variable: https://drive.google.com/file/d/16Rv3QZQvONqLI76mxFrT3usoTBRJ2Azf/view?usp=sharing

**Comment by lablabla** [comment]
I'm having the same issue, and it looks like the problem is not with `CUDA_NVRTC_LIB`, but rather that the `PYTHON_EXECUTABLE` isn't defined

I added this line to after the call the `execute_process`:

```
message(STATUS "PYTHON_EXECUTABLE=(${PYTHON_EXECUTABLE}), _retval=${_retval}. CUDA_NVRTC_LIB=${CUDA_NVRTC_LIB}")
```

and this is the output:
```
-- PYTHON_EXECUTABLE=(), _retval=No such file or directory. CUDA_NVRTC_LIB=/usr/local/cuda-11.1/lib64/libnvrtc.so
CMake Warning at torch-v1.8.0/share/cmake/Caffe2/public/cuda.cmake:199 (message):
  Failed to compute shorthash for libnvrtc.so
Call Stack (most recent call first):
  torch-v1.8.0/share/cmake/Caffe2/Caffe2Config.cmake:88 (include)
  torch-v1.8.0/share/cmake/Torch/TorchConfig.cmake:68 (find_package)
  CMakeLists.txt:5 (find_package)
```

**Comment by malfet** [comment]
@lablabla thank you for your analysis. This is exactly the source fo the problem: PYTHON_EXECUTABLE is not defined when building libtorch.
This warning is harmless, but quite distracting. Will have a fix for it soon.

**Comment by lamhoangtung** [comment]
Small bump, this still happen in libtorch 1.8.1

**Comment by michalsustr** [comment]
> Small bump, this still happen in libtorch 1.8.1

Confirming this happens in libtorch 1.8.1, using libtorch-cxx11-abi-shared-with-deps-1.8.1+cu111 binaries from the official website. Everything works, there is just the warning.

**Comment by fmigneault** [comment]
In my case, I resolved by replacing ``${PYTHON_EXECUTABLE}`` by explicit ``$ENV{PYTHON_EXECUTABLE}`` since it was defined in my env but not in cmake cache. Using pytorch 1.8.1. 

A small check of the value prior to using PYTHON_EXECUTABLE with an explicit "missing message" would be helpful. 
If it wasn't for this thread, it would have taken me a long time to guess the problem was about python, and not at all related to cuda as the error message seem to suggest.

**Comment by ttbek** [comment]
As someone pretty new to libtorch, the message can seem pretty worrisome.  I'm using the same version (1.8.1).  I am running in the Nividia CUDA docker container.  Steps to reproduce: 

1) Download the official libTorch file (libtorch-cxx11-abi-shared-with-deps-1.8.1+cu111.zip) and unpack it, as described here: 
https://pytorch.org/cppdocs/installing.html

2) Make sure your Nvidia driver is up to date and install Docker, use Docker to run a container with the appropriate versions of CUDA and CUDNN, I think the following should be ok for this version of libtorch: 

`sudo nvidia-docker run -it -v /path/with/libtorch/and/example/under/it/:/path/with/libtorch/and/example/under/it/ nvidia/cuda:11.1.1-cudnn8-devel-ubuntu20.04`
Try to continue the tutorial now that we have the prereqs in our container and we are at a shell in the container (and cd over to where we are doing the example, which should be under the volume we mounted).

3) Attempt the build, there will probably be a problem, add `export CUDA_HOME=/usr/local/cuda-11.1/` to your build script if you made one from the commands shown in the tutorial, or just run it before building.

4) The Tutorial works but you have the annoying error message this thread is about and aren't sure what it impacts, so you search the issues and find this.  You read through them but it isn't very clear where to set these, setting the environment variable like we did for `CUDA_HOME` doesn't seem to work.  fmigneault's setting of `$ENV{PYTHON_EXECUTABLE}` is where?  In `libtorch-cxx11-abi-shared-with-deps-1.8.1+cu111/libtorch/share/cmake/Caffe2/public/cuda.cmake` around line 198?  That seems like it would be an upstream change that would make just setting the env variable work like for `CUDA_HOME`, but I'm not sure that is where I want to tinker.  As a naive end user I'm partial to the approach of lamhoangtung, only that we actually need to set the CMake variable for `PYTHON_EXECUTABLE` rather than for `CUDA_NVRTC_LIB`.  

If we look back at the tutorial, https://pytorch.org/cppdocs/installing.html  you should have a file "CMakeLists.txt", in that file, add a line with the path to your python like: 
`set(PYTHON_EXECUTABLE "/usr/bin/python3")`
just above the line `find_package(Torch REQUIRED)`

You can find the path to your python by running `which python`  or `which python3` in Bash.  Of course, this container has no python, so we need to install it, e.g. `apt-get install python3`  Of course the container will have an empty cache for apt... so you actually need to `apt-get update` before you can install python. I didn't test this with python, just with python3.  

Of course this seems kind of silly since we are explicitly trying to use C++ (if we're missing with libtorch), sometimes intentionally to not use Python, so in that case we can just live with the error message?  I don't think it breaks anything on the execution side... but I don't really know what libnvirtc does or whether or not it is actually a problem or not that the shorthash fails.

**Comment by codereptile** [comment]
TLDR: This is usually caused by this error: "Command 'python' not found", you should try to install: `python-is-python3`
```
sudo apt install python-is-python3
```
Or make an alias/link for it.

**Comment by idangrady** [comment]
> TLDR: This is usually caused by this error: "Command 'python' not found", you should try to install: `python-is-python3`
> 
> ```
> sudo apt install python-is-python3
> ```
> 
> Or make an alias/link for it.

This fixed my issue!! Thanks!! :)

**Comment by snio89** [comment]
For those who suffer from this warning with libtorch 2.5.1, You have to modify `cuda.cmake`. Go to https://github.com/pytorch/pytorch/issues/129777.
